File: app/src/main/java/androidx/iot/mqtt/Mqtt.java
package androidx.iot.mqtt;

import android.content.Context;
import android.util.Log;

import org.eclipse.paho.android.service.MqttAndroidClient;
import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttCallback;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;

import java.util.concurrent.ConcurrentHashMap;

/**
 * MQTT
 */
public class Mqtt implements Imqtt, MqttCallback, IMqttActionListener {

    public final static String TAG = Mqtt.class.getSimpleName();
    /**
     * MQTT客户端
     */
    private MqttAndroidClient mqttAndroidClient;
    /**
     * MQTT参数
     */
    private MqttConnectOptions mqttConnectOptions;
    /**
     * 是否调试
     */
    private boolean debug;
    private MqttHandler handler;
    /**
     * MQTT
     */
    private static Mqtt instance;


    public Mqtt() {
        Log.i(TAG, "instantiation mqtt");
        handler = new MqttHandler();
    }

    /**
     * 是否调试
     *
     * @return
     */
    public boolean isDebug() {
        return debug;
    }

    /**
     * 设置调试模式
     *
     * @param debug
     */
    public Mqtt debug(boolean debug) {
        this.debug = debug;
        return this;
    }

    /**
     * 初始化物联网操作对象
     *
     * @param context 上下文
     * @param options 参数
     * @return
     */
    public static Mqtt initialize(Context context, MqttOptions options) {
        if (instance == null) {
            synchronized (Mqtt.class) {
                if (instance == null) {
                    instance = new Mqtt();
                }
            }
        }
        instance.setMqttAndroidClientOptions(context, options);
        return instance;
    }

    /**
     * 获取客户端对象（必须初始化才可以调用）
     *
     * @return
     */
    public static Mqtt acquire() {
        return instance;
    }

    /**
     * 设置连接参数
     *
     * @param context 上下文
     * @param options 参数
     */
    public void setMqttAndroidClientOptions(Context context, MqttOptions options) {
        mqttAndroidClient = new MqttAndroidClient(context.getApplicationContext(), options.getHost(), options.getClientId());
        mqttAndroidClient.registerResources(context.getApplicationContext());
        mqttAndroidClient.setCallback(this);
        mqttConnectOptions = new MqttConnectOptions();
        mqttConnectOptions.setUserName(options.getUserName());
        mqttConnectOptions.setPassword(options.getPassword().toCharArray());
        Log.i(TAG, "service url " + options.getHost());
        if (debug) {
            Log.i(TAG, options.getClientId() + " " + options.getUserName() + " " + options.getPassword());
        }
    }

    /**
     * 设置Android Mqtt客户端
     *
     * @param mqttAndroidClient
     */
    public void setMqttAndroidClient(MqttAndroidClient mqttAndroidClient) {
        if (mqttAndroidClient != null) {
            mqttAndroidClient.setCallback(this);
        }
        this.mqttAndroidClient = mqttAndroidClient;
    }

    /**
     * Android Mqtt客户端
     *
     * @return
     */
    public MqttAndroidClient getMqttAndroidClient() {
        return mqttAndroidClient;
    }


    /**
     * 设置Mqtt参数
     *
     * @param mqttConnectOptions
     */
    public void setMqttConnectOptions(MqttConnectOptions mqttConnectOptions) {
        this.mqttConnectOptions = mqttConnectOptions;
    }

    /**
     * Mqtt连接参数
     *
     * @return
     */
    public MqttConnectOptions getMqttConnectOptions() {
        return mqttConnectOptions;
    }

    @Override
    public Imqtt connect() {
        try {
            if (mqttAndroidClient != null) {
                mqttAndroidClient.connect(mqttConnectOptions, null, this);
            }
        } catch (MqttException e) {
            e.printStackTrace();
        }
        return this;
    }

    @Override
    public void disconnect() {
        try {
            if (mqttAndroidClient != null) {
                mqttAndroidClient.unregisterResources();
                Thread.sleep(50);
                mqttAndroidClient.disconnect();
            }
        } catch (MqttException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean isConnected() {
        if (mqttAndroidClient == null) {
            return false;
        }
        return mqttAndroidClient.isConnected();
    }

    //*******************状态**********************
    @Override
    public void connectionLost(Throwable cause) {
        Log.i(TAG, "connection lost");
        try {
            if (mqttAndroidClient != null) {
                mqttAndroidClient.connect();
            }
        } catch (MqttException e) {
            e.printStackTrace();
        }
        if (connectHashMap != null) {
            for (Long key : connectHashMap.keySet()) {
                handler.connectionLost(cause, connectHashMap.get(key));
            }
        }
    }

    @Override
    public void messageArrived(String topic, MqttMessage message) {
        String payload = new String(message.getPayload());
        if (debug) {
            Log.i(TAG, "received " + topic + " " + payload);
        }
        if (messageHashMap != null) {
            for (Long key : messageHashMap.keySet()) {
                handler.messageArrived(topic, message, messageHashMap.get(key));
            }
        }
    }

    @Override
    public void deliveryComplete(IMqttDeliveryToken token) {
        if (messageHashMap != null) {
            for (Long key : messageHashMap.keySet()) {
                handler.deliveryComplete(token, messageHashMap.get(key));
            }
        }
    }

    //*********************[连接成功]**********************

    @Override
    public void onSuccess(IMqttToken token) {
        Log.i(TAG, "connection successful");
        if (connectHashMap != null) {
            for (Long key : connectHashMap.keySet()) {
                handler.connectionSuccessful(token, connectHashMap.get(key));
            }
        }
    }

    @Override
    public void onFailure(IMqttToken token, Throwable exception) {
        Log.i(TAG, "connection failed");
        if (connectHashMap != null) {
            for (Long key : connectHashMap.keySet()) {
                handler.connectFailed(token, exception, connectHashMap.get(key));
            }
        }
    }

    //***********************[END]*****************************

    @Override
    public Imqtt publish(String topic, String payload, IMqttActionListener listener) {
        try {
            if (mqttAndroidClient != null) {
                if (mqttAndroidClient.isConnected() == false) {
                    mqttAndroidClient.connect();
                }
                MqttMessage message = new MqttMessage();
                int id = (int) System.currentTimeMillis() + 200;
                message.setPayload(payload.getBytes());
                message.setId(id);
                message.setQos(0);
                mqttAndroidClient.publish(topic, message, null, listener);
                if (debug) {
                    Log.i(TAG, "publish id:" + id + " " + topic + " " + payload);
                }
            }
        } catch (MqttException e) {
            e.printStackTrace();
        }
        return this;
    }

    @Override
    public Imqtt publish(String topic, String payload) {
        return publish(topic, payload, null);
    }

    @Override
    public Imqtt subscribe(String topic, IMqttActionListener listener) {
        try {
            if (debug) {
                Log.i(TAG, "subscribe " + topic);
            }
            mqttAndroidClient.subscribe(topic, 0, null, listener);
        } catch (MqttException e) {
            e.printStackTrace();
        }
        return this;
    }

    @Override
    public Imqtt subscribe(String topic) {
        return subscribe(topic, null);
    }

    /**
     * 信息监听Map
     */
    private ConcurrentHashMap<Long, OnMessageListener> messageHashMap;

    @Override
    public long addMessageListener(OnMessageListener listener) {
        if (messageHashMap == null) {
            messageHashMap = new ConcurrentHashMap<>();
        }
        long mid = System.currentTimeMillis() + messageHashMap.size() + 1;
        messageHashMap.put(mid, listener);
        return mid;
    }

    /**
     * 连接监听Map
     */
    private ConcurrentHashMap<Long, OnConnectListener> connectHashMap;

    @Override
    public long addConnectListener(OnConnectListener listener) {
        if (connectHashMap == null) {
            connectHashMap = new ConcurrentHashMap<>();
        }
        long cid = System.currentTimeMillis() + connectHashMap.size() + 1;
        connectHashMap.put(cid, listener);
        return cid;
    }

    @Override
    public Imqtt remove(long... ids) {
        for (long id : ids) {
            if (messageHashMap != null) {
                messageHashMap.remove(id);
            }
            if (connectHashMap != null) {
                connectHashMap.remove(id);
            }
        }
        return this;
    }

    @Override
    public Imqtt clear() {
        if (messageHashMap != null) {
            messageHashMap.clear();
        }
        if (connectHashMap != null) {
            connectHashMap.clear();
        }
        return this;
    }

    @Override
    public void destroy() {
        disconnect();
        if (messageHashMap != null) {
            messageHashMap.clear();
            connectHashMap = null;
        }
        if (connectHashMap != null) {
            connectHashMap.clear();
            connectHashMap = null;
        }
        if (handler != null) {
            handler.destroy();
            handler = null;
        }
        instance = null;
    }

}

File: src/main/java/androidx/iot/mqtt/Mqtt.kt
package androidx.iot.mqtt

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.util.Log
import androidx.iot.link.AuthType
import org.eclipse.paho.android.service.MqttAndroidClient
import org.eclipse.paho.client.mqttv3.IMqttActionListener
import org.eclipse.paho.client.mqttv3.MqttCallback
import org.eclipse.paho.client.mqttv3.MqttConnectOptions
import org.eclipse.paho.client.mqttv3.MqttMessage
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence

/**
 * MQTT连接基础类
 */
class Mqtt(context: Context, options: Options, callback: MqttCallback?) {

    private val TAG = "Mqtt"

    /**
     * 客户端
     */
    private var client: MqttAndroidClient? = null

    /**
     * MQTT连接参数
     */
    private var connOpts: MqttConnectOptions

    /**
     * 初始化
     */
    init {
        val applicationContext = context.applicationContext
        val isNetworkAvailable = isNetworkAvailable(applicationContext)
        when (options.authType) {
            AuthType.CONNECT, AuthType.CONNWL -> {
                client = MqttAndroidClient(applicationContext, options.host, options.clientId)
                if (isNetworkAvailable){
                    client?.registerResources(applicationContext)
                }
                client?.setCallback(callback)
                connOpts = MqttConnectOptions()
                connOpts.connectionTimeout = 15
                connOpts.keepAliveInterval = 60
                connOpts.userName = options.username
                connOpts.password = options.password.toCharArray()
                connOpts.isAutomaticReconnect = true
                connOpts.isCleanSession = false
            }

            AuthType.REGISTER, AuthType.REGNWL -> {
                client = MqttAndroidClient(
                    applicationContext,
                    options.host,
                    options.clientId,
                    MemoryPersistence()
                )
                if (isNetworkAvailable){
                    client?.registerResources(applicationContext)
                }
                client?.setCallback(callback)
                connOpts = MqttConnectOptions()
                connOpts.connectionTimeout = 15
                connOpts.mqttVersion = 4// MQTT 3.1.1
                connOpts.userName = options.username
                connOpts.password = options.password.toCharArray()
                connOpts.isAutomaticReconnect = false//MQTT动态注册协议规定必须关闭自动重连
            }
        }
    }

    /**
     * 是否已连接
     * @return 是否连接
     */
    fun isConnected(): Boolean {
        return client?.isConnected == true
    }

    /**
     * 连接服务端
     * @param callback 连接监听
     */
    fun connect(context: Context,callback: IMqttActionListener?) {
        if (isNetworkAvailable(context)){
            client?.connect(connOpts, context, callback)
        }
    }

    /**
     * 向 MQTT 服务器发布消息
     * @param topic 指定消息要发布到的 MQTT 主题
     * @param payload 消息的实际内容，以字节数组的形式表示
     * @param qos 指定消息的服务质量等级
     *      0：最多一次（At most once）。消息可能会丢失，不会进行重传，消息只发送一次。这是最快但不可靠的服务质量等级。
     *      1：至少一次（At least once）。消息至少会被传递一次，如果发送失败会进行重传，确保消息最终到达接收方，但可能会有重复消息。
     *      2：恰好一次（Exactly once）。消息只会被传递一次，不会丢失也不会重复，通过复杂的握手协议来保证消息的准确性，但性能相对较低。
     * @param retained 指定消息是否为保留消息。如果设置为 true，MQTT 服务器会将该消息保留，并在新的客户端订阅该主题时立即将保留消息发送给客户端。如果设置为 false，则消息不会被保留。
     * @param userContext 用户自定义的上下文对象，可以是任何类型的对象
     * @param callback 一个实现了 IMqttActionListener 接口的对象，用于异步处理消息发布的结果
     */
    fun public(
        topic: String,
        payload: ByteArray,
        qos: Int,
        retained: Boolean,
        userContext: Any?,
        callback: IMqttActionListener?
    ) {
        client?.publish(topic, payload, qos, retained, userContext, callback)
    }

    /**
     * 向 MQTT 服务器发布消息
     * @param topic 指定消息要发布到的 MQTT 主题
     * @param payload 消息的实际内容，以字节数组的形式表示
     * @param qos 指定消息的服务质量等级
     *      0：最多一次（At most once）。消息可能会丢失，不会进行重传，消息只发送一次。这是最快但不可靠的服务质量等级。
     *      1：至少一次（At least once）。消息至少会被传递一次，如果发送失败会进行重传，确保消息最终到达接收方，但可能会有重复消息。
     *      2：恰好一次（Exactly once）。消息只会被传递一次，不会丢失也不会重复，通过复杂的握手协议来保证消息的准确性，但性能相对较低。
     * @param retained 指定消息是否为保留消息。如果设置为 true，MQTT 服务器会将该消息保留，并在新的客户端订阅该主题时立即将保留消息发送给客户端。如果设置为 false，则消息不会被保留。
     */
    fun public(topic: String, payload: ByteArray, qos: Int, retained: Boolean) {
        client?.publish(topic, payload, qos, retained)
    }

    /**
     * 向 MQTT 服务器发布消息
     * @param topic 指定消息要发布到的 MQTT 主题
     * @param payload 消息的实际内容，以字节数组的形式表示
     */
    fun publish(topic: String, payload: String) {
        client?.publish(topic, payload.toByteArray(), 0, false)
    }

    /**
     * 订阅 MQTT 主题
     * @param topic 数组中的每个元素代表一个要订阅的 MQTT 主题
     * @param qos 服务质量等级数组，与 topic 数组一一对应
     * @param userContext 用户自定义的上下文对象，可以是任意类型的对象
     * @param callback 用于异步处理订阅操作的结果
     */
    fun subscribe(
        topic: Array<String>,
        qos: IntArray,
        userContext: Any?,
        callback: IMqttActionListener?
    ) {
        client?.subscribe(topic, qos, userContext, callback)
    }

    /**
     * 订阅 MQTT 主题
     * @param topic 数组中的每个元素代表一个要订阅的 MQTT 主题
     */
    fun subscribe(topic: Array<String>) {
        client?.subscribe(topic, IntArray(topic.size) { 0 })
    }

    /**
     * 订阅 MQTT 主题
     * @param topic 要订阅的 MQTT 主题名称
     * @param qos 指定订阅该主题的服务质量等级
     *      0：最多一次（At most once），消息可能会丢失，不保证消息一定能到达。
     *      1：至少一次（At least once），消息至少会被传递一次，如果发送失败会进行重传，但可能会有重复消息。
     *      2：恰好一次（Exactly once），消息只会被传递一次，不会丢失也不会重复，通过复杂的握手协议来保证消息的准确性。
     * @param userContext 用户自定义的上下文对象，可以是任意类型的对象
     * @param callback 用于异步处理订阅操作的结果
     */
    fun subscribe(topic: String, qos: Int, userContext: Any?, callback: IMqttActionListener?) {
        client?.subscribe(topic, qos, userContext, callback)
    }

    /**
     * 订阅 MQTT 主题
     * @param topic 要订阅的 MQTT 主题名称
     * @param qos 指定订阅该主题的服务质量等级
     *      0：最多一次（At most once），消息可能会丢失，不保证消息一定能到达。
     *      1：至少一次（At least once），消息至少会被传递一次，如果发送失败会进行重传，但可能会有重复消息。
     *      2：恰好一次（Exactly once），消息只会被传递一次，不会丢失也不会重复，通过复杂的握手协议来保证消息的准确性。
     */
    fun subscribe(topic: String, qos: Int) {
        client?.subscribe(topic, qos)
    }

    /**
     * 取消订阅指定主题
     * @param topic 要订阅的 MQTT 主题名称数组
     * @param userContext 用户自定义的上下文对象，可以是任意类型的对象
     * @param callback 用于异步处理订阅操作的结果
     */
    fun unsubscribe(topic: Array<String>, userContext: Any?, callback: IMqttActionListener?) {
        client?.unsubscribe(topic, userContext, callback)
    }

    /**
     * 取消订阅指定主题
     * @param topic 要订阅的 MQTT 主题名称数组
     */
    fun unsubscribe(topic: Array<String>) {
        client?.unsubscribe(topic)
    }

    /**
     * 取消订阅指定主题
     * @param topic 要订阅的 MQTT 主题名称
     * @param userContext 用户自定义的上下文对象，可以是任意类型的对象
     * @param callback 用于异步处理订阅操作的结果
     */
    fun unsubscribe(topic: String, userContext: Any?, callback: IMqttActionListener?) {
        client?.unsubscribe(topic, userContext, callback)
    }

    /**
     * 取消订阅指定主题
     * @param topic 要订阅的 MQTT 主题名称
     */
    fun unsubscribe(topic: String) {
        client?.unsubscribe(topic)
    }

    /**
     * 断开连接
     */
    fun disconnect() {
        Log.d(TAG, "disconnect")
        client?.let {
            it.unregisterResources()
            it.disconnect()
        }
    }

    /**
     * 当前网络是否可用
     */
   private fun isNetworkAvailable(context: Context): Boolean {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        // 对于 Android 10（API 级别 29）及以上版本
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            val network = connectivityManager.activeNetwork
            if (network != null) {
                val networkCapabilities = connectivityManager.getNetworkCapabilities(network)
                return networkCapabilities != null && (
                        networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
                                networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) ||
                                networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)
                        )
            }
        }
        // 对于 Android 5.0（API 级别 21）到 Android 9（API 级别 28）
        else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val networks = connectivityManager.allNetworks
            for (net in networks) {
                val networkInfo = connectivityManager.getNetworkInfo(net)
                if (networkInfo != null && networkInfo.isConnected) {
                    return true
                }
            }
        }
        // 对于 Android 5.0 以下版本
        else {
            val networkInfo = connectivityManager.activeNetworkInfo
            return networkInfo != null && networkInfo.isConnected
        }
        return false
    }


}
